### PR TITLE
Add grape methods for classifiers

### LINKED ISSUE #77: Set up methods for using Ensmallen/Embiggen native methods
We support TF methods for link prediction through [mlp_model.py](https://github.com/Knowledge-Graph-Hub/NEAT/blob/main/neat/link_prediction/mlp_model.py), but other methods previously only accessible through TF are increasingly available through Ensmallen/Embiggen. We should support these methods, too.

### LINKED ISSUE #85: Convert edge prediction methods to grape methods
The embedding methods are updated to be compatible with the latest `grape` in #82, but this doesn't include updates for edge prediction. We'd like to update those, too.

### PR DESCRIPTION
**Pull request by caufieldjh**

### CODE REVIEW

**Comment by caufieldjh** [comment]
The parts appear to work on their own (i.e., can specify an Ensmallen edge prediction model and can apply it to a graph) but running a full config with a perceptron (`grape.edge_prediction.PerceptronEdgePrediction `) yields the following:
```
$ neat run --config giraffe.yaml 
  0%|                                                                                                                                       | 0/1 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "/home/harry/neat-env/bin/neat", line 8, in <module>
    sys.exit(cli())
  File "/home/harry/neat-env/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/harry/neat-env/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/harry/neat-env/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/harry/neat-env/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/harry/neat-env/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/harry/neat-env/lib/python3.8/site-packages/neat_ml/cli.py", line 78, in run
    raise NotImplementedError(f"{model} isn't implemented yet")
NotImplementedError: None isn't implemented yet
```

My guess is that the handoff from embedder to classifier isn't passing the model object, it's just trying to get the name of it first.

**Comment by caufieldjh** [comment]
That was just an issue of the CLI not recognizing the model name.

Now we're here:
```
Traceback (most recent call last):
  File "/home/harry/neat-env/bin/neat", line 8, in <module>
    sys.exit(cli())
  File "/home/harry/neat-env/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/harry/neat-env/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/harry/neat-env/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/harry/neat-env/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/harry/neat-env/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/harry/neat-env/lib/python3.8/site-packages/neat_ml/cli.py", line 107, in run
    np.around(model.predict(validation_data[0]), decimals=0)
  File "/home/harry/neat-env/lib/python3.8/site-packages/neat_ml/link_prediction/model.py", line 46, in predict
    return self.model.predict(predict_data)  # type: ignore
  File "/home/harry/neat-env/lib/python3.8/site-packages/embiggen/edge_prediction/edge_prediction_model.py", line 323, in predict
    predictions = super().predict(
  File "/home/harry/neat-env/lib/python3.8/site-packages/embiggen/utils/abstract_models/abstract_classifier_model.py", line 869, in predict
    if not graph.has_nodes():
AttributeError: 'numpy.ndarray' object has no attribute 'has_nodes'
```

**Comment by caufieldjh** [comment]
Classifier building and validation works now with graph + embedding + validation, but fails when trying to apply the classifier as it can't load it from a file (since we can't save the model).

**Comment by caufieldjh** [comment]
Just need to:

- [ ] Ensure grape models don't output existing edges in predictions, if asked
- [ ] Ensure node type filter is used for grape models
- [ ] Finish cleanup/linting for flake8

**Comment by caufieldjh** [comment]
Re: the gen_src_dist_pair function, the current behavior is essentially to generate all potential edges (except, in practice, we don't, because that would require more memory than we're likely to have), add them to a list (that's the problem step, really), and pass that back to the predict_links function to check if they match the node type filter. We use this list to generate the corresponding edge embeddings with edge_embedding_for_predict. 

For the `grape` models it may make more sense to build the model, run predict_proba, then remove all existing edges from the results without trying to generate every potential combination first (the link prediction subsampling already happens in grape, AFAIK).

So that refactor can go in a different PR.

File: neat_ml/cli.py [hunk 1]
```diff
@@ -66,14 +69,22 @@ def run(config: str) -> None:
...
+            gmodels = [mname.lower() for mname in gmodels]
             if classifier["classifier_name"].lower() == "neural network":
                 model = MLPModel(classifier, outdir=yhelp.outdir())
-            elif classifier["classifier_name"].lower() in [
-                "decision tree",
-                "logistic regression",
-                "random forest",
-            ]:
+            elif (
+                classifier["classifier_name"].lower() == "logistic regression"
```
Comment (hrshdhgd) [inline neat_ml/cli.py:79]: Do we want to create a separate section of constants at the top (or constants.py) where:
```
LOGISTIC_REGRESSION = "logistic regression"
```
and then apply here?

File: neat_ml/cli.py [hunk 2]
```diff
@@ -66,14 +69,22 @@ def run(config: str) -> None:
...
+            ):
                 model = SklearnModel(classifier, outdir=yhelp.outdir())
+            elif (
+                classifier["classifier_name"].lower() in gmodels
+                or classifier["classifier_name"].lower() + " classifier"
```
Comment (hrshdhgd) [inline neat_ml/cli.py:84]: Similar constant declaration for "classifier" , maybe?

File: neat_ml/cli.py [hunk 3]
```diff
@@ -86,15 +97,46 @@ def run(config: str) -> None:
...
+                    (yhelp.embedding_outfile()), index_col=0, header=None
+                )
+                history_obj = model.fit(  # type: ignore
+                    graph=graph_obj, node_features=embed_obj  # type: ignore
+                )
+
+            if type(model) == GrapeModel:
+                if "pos_edge_filepath" in yhelp.val_graph_args():
+                    val_graph_obj = Graph.from_csv(
+                        node_path=yhelp.main_graph_args()["node_path"],
```
Comment (hrshdhgd) [inline neat_ml/cli.py:115]: Constant for "node_path"?

File: neat_ml/cli.py [hunk 4]
```diff
@@ -86,15 +97,46 @@ def run(config: str) -> None:
...
+                        edge_path=yhelp.val_graph_args()["pos_edge_filepath"],
```
Comment (hrshdhgd) [inline neat_ml/cli.py:116]: Constant for all slots?

File: neat_ml/cli.py [hunk 5]
```diff
@@ -86,15 +97,46 @@ def run(config: str) -> None:
...
+                        nodes_column=yhelp.main_graph_args()["nodes_column"],
```
Comment (hrshdhgd) [inline neat_ml/cli.py:117]: Constant for all slots?

File: neat_ml/cli.py [hunk 6]
```diff
@@ -86,15 +97,46 @@ def run(config: str) -> None:
...
+                        node_list_node_types_column=yhelp.main_graph_args()[
+                            "node_list_node_types_column"
```
Comment (hrshdhgd) [inline neat_ml/cli.py:119]: Constant for all slots?

File: neat_ml/cli.py [hunk 7]
```diff
@@ -111,7 +153,13 @@ def run(config: str) -> None:
     if yhelp.do_apply_classifier():
         # take graph, classifier, biolink node types and cutoff
         for clsfr_id in yhelp.get_classifier_id_for_prediction():
-            classifier_kwargs = yhelp.make_classifier_args(clsfr_id)
+            classifier = yhelp.get_classifier_from_id(clsfr_id)
+            if classifier["classifier_type"].startswith("grape"):
```
Comment (hrshdhgd) [inline neat_ml/cli.py:157]: Constant for all slots?

File: neat_ml/cli.py (hunk 1 above)
Comment (caufieldjh) [reply neat_ml/cli.py:79]: Good idea - could even be a list of acceptable variations/synonyms

File: neat_ml/yaml_helper/yaml_helper.py [hunk 8]
```diff
@@ -535,6 +521,10 @@ def make_classifier_args(self, cl_id: str) -> dict:
                 ),
             )
 
+        elif model_in is not None:
+            # Workaround for grape save/load not implemented yet
+            # We just take the model object as an argument
+            classifier_args_dict["model"] = model_in
```
Comment (hrshdhgd) [inline neat_ml/yaml_helper/yaml_helper.py:527]: Constant?

**Comment by hrshdhgd** [comment]
Apologies in advance for the annoying comments about constants, I bring this up because we do this in `sssom` and it is really convenient and is easy if refactors are needed.

**Comment by caufieldjh** [comment]
> Apologies in advance for the annoying comments about constants, I bring this up because we do this in `sssom` and it is really convenient and is easy if refactors are needed.

No worries, it's probably time for those to be constants. Since these are all defined in the neat schema, we could use the schema to define the set of constants

**Comment by caufieldjh** [comment]
Tests failing due to some new issue between `tox` and `flake8`

**Comment by hrshdhgd** [comment]
For now I've implemented an unrecommended fix of commenting out the troublemaker `flake8` plugins:
```
flake8-bandit
flake8-isort
```
`flake8` had a new release go out today and that may have some influence over this issue.

**Comment by caufieldjh** [comment]
> Another solution: I brought back the plugins and limited `flake8 < 5.0.0`. This works too.

Excellent - we may just have to keep it like that for a bit

**Pull request merged**